File: frontend/src/api/recipes.ts
import api from './index'
import type { RecipeListItem, RecipeDetail } from '@/types'

export interface RecipeListParams {
  category?: string
  cuisine_type?: string
  difficulty?: string
  max_prep_time?: number
  tags?: string
  page?: number
  page_size?: number
}

export interface RecipeListResponse {
  total: number
  page: number
  page_size: number
  items: RecipeListItem[]
}

export async function getRecipes(params?: RecipeListParams): Promise<RecipeListResponse> {
  const { data } = await api.get('/recipes', { params })
  return {
    total: Number(data?.total) || 0,
    page: Number(data?.page) || 1,
    page_size: Number(data?.page_size) || 20,
    items: Array.isArray(data?.items) ? data.items.map((item: RecipeListItem) => ({
      ...item,
      recipe_id: Number(item.recipe_id),
      prep_time: Number(item.prep_time) || 0,
      cook_time: Number(item.cook_time) || 0,
      total_calories: Number(item.total_calories) || 0,
      tags: Array.isArray(item.tags) ? item.tags.map(String) : [],
    })) : [],
  }
}

export async function getRecipeDetail(id: number): Promise<RecipeDetail> {
  const { data } = await api.get(`/recipes/${id}`)
  return {
    ...data,
    recipe_id: Number(data.recipe_id),
    description: String(data.description || ''),
    prep_time: Number(data.prep_time) || 0,
    cook_time: Number(data.cook_time) || 0,
    servings: Math.max(1, Number(data.servings) || 1),
    total_calories: Number(data.total_calories ?? data.nutrition?.calories) || 0,
    estimated_cost: Number(data.estimated_cost) || 0,
    tags: Array.isArray(data.tags) ? data.tags.map(String) : [],
    steps: Array.isArray(data.steps) ? data.steps.map((step: { step?: unknown; content?: unknown }, index: number) => ({
      step: Number(step.step) || index + 1,
      content: String(step.content || ''),
    })) : [],
    ingredients: Array.isArray(data.ingredients) ? data.ingredients.map((item: RecipeDetail['ingredients'][number]) => ({
      ...item,
      ingredient_id: Number(item.ingredient_id),
      quantity: Number(item.quantity) || 0,
      is_optional: Boolean(item.is_optional),
    })) : [],
    nutrition: data.nutrition ? {
      calories: Number(data.nutrition.calories) || 0,
      protein: Number(data.nutrition.protein) || 0,
      fat: Number(data.nutrition.fat) || 0,
      carbs: Number(data.nutrition.carbs) || 0,
      fiber: Number(data.nutrition.fiber) || 0,
    } : null,
  }
}

File: backend/app/api/routes/recipes.py
"""菜谱 API 路由"""

from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.database import get_db
from app.api.schemas.recipe import (
    RecipeListItem, RecipeDetail, RecipeListResponse,
    RecipeIngredientItem, RecipeStep,
)
from app.services.recipe_service import (
    get_recipe, get_recipes, get_recipe_ingredients,
)

router = APIRouter(tags=["recipes"])


@router.get("/recipes", response_model=RecipeListResponse)
def api_get_recipes(
    category: Optional[str] = None,
    cuisine_type: Optional[str] = None,
    difficulty: Optional[str] = None,
    max_prep_time: Optional[int] = None,
    tags: Optional[str] = Query(None, description="逗号分隔的标签"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=50),
    db: Session = Depends(get_db),
):
    tag_list = tags.split(",") if tags else None
    items, total = get_recipes(
        db,
        category=category,
        cuisine_type=cuisine_type,
        difficulty=difficulty,
        max_prep_time=max_prep_time,
        tags=tag_list,
        page=page,
        page_size=page_size,
    )
    return {
        "total": total,
        "page": page,
        "page_size": page_size,
        "items": [
            RecipeListItem(
                recipe_id=r.recipe_id,
                name=r.name,
                category=r.category,
                cuisine_type=r.cuisine_type,
                difficulty=r.difficulty,
                prep_time=r.prep_time,
                cook_time=r.cook_time,
                image_url=r.image_url,
                total_calories=float(r.total_calories),
                tags=r.tags,
            )
            for r in items
        ],
    }


@router.get("/recipes/{recipe_id}", response_model=RecipeDetail)
def api_get_recipe(recipe_id: int, db: Session = Depends(get_db)):
    recipe = get_recipe(db, recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="菜谱不存在")

    ingredients_raw = get_recipe_ingredients(db, recipe_id)
    total_cost = sum(i["estimated_cost"] for i in ingredients_raw)

    ingredients = [
        RecipeIngredientItem(
            ingredient_id=i["ingredient_id"],
            name=i["name"],
            quantity=i["quantity"],
            unit=i["unit"],
            is_optional=i["is_optional"],
        )
        for i in ingredients_raw
    ]

    # 从食材营养汇总计算菜谱营养
    nutrition = {
        "calories": float(recipe.total_calories),
        "protein": float(recipe.total_protein),
        "fat": float(recipe.total_fat),
        "carbs": float(recipe.total_carbs),
        "fiber": float(recipe.total_fiber),
    }

    return RecipeDetail(
        recipe_id=recipe.recipe_id,
        name=recipe.name,
        description=recipe.description,
        category=recipe.category,
        cuisine_type=recipe.cuisine_type,
        difficulty=recipe.difficulty,
        prep_time=recipe.prep_time,
        cook_time=recipe.cook_time,
        servings=recipe.servings,
        steps=[RecipeStep(**s) for s in (recipe.steps or [])],
        image_url=recipe.image_url,
        ingredients=ingredients,
        nutrition=nutrition,
        estimated_cost=round(total_cost, 2),
        tags=recipe.tags,
    )


@router.get("/recipes/{recipe_id}/nutrition")
def api_get_recipe_nutrition(recipe_id: int, db: Session = Depends(get_db)):
    recipe = get_recipe(db, recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="菜谱不存在")

    ingredients_raw = get_recipe_ingredients(db, recipe_id)
    total_cost = sum(i["estimated_cost"] for i in ingredients_raw)

    return {
        "recipe_id": recipe_id,
        "name": recipe.name,
        "nutrition": {
            "calories": float(recipe.total_calories),
            "protein": float(recipe.total_protein),
            "fat": float(recipe.total_fat),
            "carbs": float(recipe.total_carbs),
            "fiber": float(recipe.total_fiber),
        },
        "ingredients": [
            {
                "name": i["name"],
                "quantity": i["quantity"],
                "unit": i["unit"],
                "estimated_cost": i["estimated_cost"],
                "nutrition_per_100g": i["nutrition"],
            }
            for i in ingredients_raw
        ],
        "total_estimated_cost": round(total_cost, 2),
    }
